src/GovUk.Education.ExploreEducationStatistics.Notifier/Functions/ReleaseNotifier.cs

```diff
@@ -0,0 +1,184 @@
+            // service allows multiple
+            foreach (var supersededPublication in msg.SupersededPublications)
+            {
+                var supersededPublicationId = supersededPublication.Id;
+                var supersededPublicationTitle = supersededPublication.Title;
+
+                var query = new TableQuery<SubscriptionEntity>()
+                    .Where(TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal,
+                        supersededPublicationId.ToString()));
+
+                logger.LogInformation("Emailing subscribers from a superseded publication");
+                var numSupersedeSubscriberEmailsSent = 0;
+
+                TableContinuationToken? token = null;
+                do
+                {
+                    var resultSegment =
+                        await table.ExecuteQuerySegmentedAsync(query, token);
+                    token = resultSegment.ContinuationToken;
+
+                    numSupersedeSubscriberEmailsSent += resultSegment.Results.Count;
+                    foreach (var entity in resultSegment.Results)
+                    {
+                        var email = entity.RowKey;
+
+                        if (sentToEmailAddresses.Contains(email))
+                        {
+                            numSupersedeSubscriberEmailsSent--;
+                            continue;
+                        }
+                        sentToEmailAddresses.Add(email);
+
+                        var unsubscribeToken =
+                            _tokenService.GenerateToken(tokenSecretKey, email, DateTime.UtcNow.AddYears(1));
+                        var values = new Dictionary<string, dynamic>
+                        {
+                            // Use values from the queue just in case the name or slug of the publication changes
+                            { "publication_name", msg.PublicationName },
+                            { "release_name", msg.ReleaseName },
+                            {
+                                "release_link",
+                                $"{webApplicationBaseUrl}find-statistics/{msg.PublicationSlug}/{msg.ReleaseSlug}"
+                            },
+                            {
+                                // NOTE: update_note not used by original release email template
+                                "update_note", msg.UpdateNote
+                            },
+                            { "unsubscribe_link", $"{baseUrl}{supersededPublicationId}/unsubscribe/{unsubscribeToken}" },
+                            { "superseded_publication_title", supersededPublicationTitle },
+                        };
+
+                        _emailService.SendEmail(notificationClient, email, newReleaseSupersededSubscribersTemplateId, values);
+                    }
+                } while (token != null);
+
+                logger.LogInformation("Emailed {NumEmailsSent} subscribers from a superseded publication",
```

Comment:
Just leaving a minor comment as I was thinking in general about future improvements to monitoring: Would be great to alert / trend data based on either:
*  expectations were thrown on failures around sending notifications (from the EES notifier side)
* upstream errors calling the notify service happened (i.e. the gov uk notify client was in the middle of an incident and failed to deal with EES's API request)

Reply:
I think Ben has raised before that we currently don't wait for a response from Notify on whether the email was sent successfully - we just trust it to do so. Although I just looked if we had a ticket for it - I suspected we would - but couldn't find it.

A lot of this Notifier code just hasn't been touched in ages - and I think that is because despite it's problems, it's worked well enough, so improving it hasn't been a priority.

Reply:
There is now a ticket to pretty much rewrite the Notifier project - although you cannot see it as you're not working on this project anymore :stuck_out_tongue: